embassy-mspm0/src/watchdog.rs

```diff
@@ -0,0 +1,330 @@
+        match self {
+            //  divide by 1
+            Self::USec1950
+            | Self::USec7810
+            | Self::USec31250
+            | Self::MSec130
+            | Self::Sec1
+            | Self::Sec8
+            | Self::Sec64
+            | Self::Sec1024 => 0u8,
+            //  divide by 2
+            Self::USec3910
+            | Self::USec15630
+            | Self::USec62500
+            | Self::MSec250
+            | Self::Sec2
+            | Self::Sec16
+            | Self::Sec128
+            | Self::Sec2048 => 1u8,
+            //  divide by 3
+            Self::USec5860
+            | Self::USec23440
+            | Self::USec93750
+            | Self::MSec380
+            | Self::Sec3
+            | Self::Sec24
+            | Self::Sec192
+            | Self::Sec3072 => 2u8,
+            //  divide by 4
+            Self::USec32250
+            | Self::USec125000
+            | Self::MSec500
+            | Self::Sec4
+            | Self::Sec32
+            | Self::Sec256
+            | Self::Sec4096 => 3u8,
+            //  divide by 5
+            Self::USec9770
+            | Self::USec39060
+            | Self::USec156250
+            | Self::MSec630
+            | Self::Sec5
+            | Self::Sec40
+            | Self::Sec320
+            | Self::Sec5120 => 4u8,
+            //  divide by 6
+            Self::USec11720
+            | Self::USec46880
+            | Self::USec187500
+            | Self::MSec750
+            | Self::Sec6
+            | Self::Sec48
+            | Self::Sec384
+            | Self::Sec6144 => 5u8,
+            //  divide by 7
+            Self::USec13670
+            | Self::USec54690
+            | Self::USec218750
+            | Self::MSec880
+            | Self::Sec7
+            | Self::Sec56
+            | Self::Sec448
+            | Self::Sec7168 => 6u8,
+            //  divide by 8
+            Self::Sec512 | Self::Sec8192 => 7u8,
+        }
+    }
+}
+
+/// Timeout percentage that is treated as "too early" and generates violation.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ClosedWindowPercentage {
+    // window period is not used
+    Zero,
+    // 12.5% percents
+    Twelve,
+    // 18.75% percents
+    Eighteen,
+    // 25% percents
+    TwentyFive,
+    // 50% percents
+    Fifty,
+    // 75% percents
+    SeventyFive,
+    // 81.25% percents
+    EightyOne,
+    // 87.5% percents
+    EightySeven,
+}
+
+#[non_exhaustive]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+/// Watchdog Config
+pub struct Config {
+    /// Watchdog timeout
+    pub timeout: Timeout,
+
+    /// closed window percentage
+    pub closed_window: ClosedWindowPercentage,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            timeout: Timeout::Sec1,
+            closed_window: ClosedWindowPercentage::Zero,
+        }
+    }
+}
+
+pub struct Watchdog<'d, T: Instance> {
```

Comment:
Given some chips have multiple watchdog instances, do we want to consider erasing the generic type here?

You would only need to store a `wwdt::Wwdt` inside which is 4 bytes total (its a pointer pretty much).

Reply:
just replacing  generic type with `wwdt::Wwdt ` complains about missing trait `PeripheralType` for `Wwdt`. 

I guess it would require some some macro magic, any hints to fix it?

Reply:
what we receive in `new` function, is an instance of generated structs  (generated by `embassy-hal-internal/src/macros.rs` ) with `PeripheralType` trait, not `wwdt::Wwdt`. And in `impl_wwdt_instance!` we map one with another to get to access to the registers.

But going back to your comment about "multiple watchdog instances", I adapted the example with WWDT0 & WWDT1 on launchpad g3507. it works with the current implementation well